File: getGridID.m
function [gridID, neighour_list] = getGridID(point, grid_size, x_lower, x_upper, y_lower, y_upper, z_lower, z_upper,dimension,find_neighbor)
    % point: 3D point coordinates [x, y, z]
    % grid_size: size of each grid in the discretized 3D space
    % x_lower, x_upper: lower and upper bounds for the x dimension
    % y_lower, y_upper: lower and upper bounds for the y dimension
    % z_lower, z_upper: lower and upper bounds for the z dimension

    neighour_list = [];


    if dimension == 3
        if point(3) < 0
            point(3) = 0;
        end
    % Calculate grid indices in each dimension
    x_size = (x_upper - x_lower) / grid_size;
    y_size = (y_upper - y_lower) / grid_size;
    z_size = (z_upper - z_lower) / grid_size;
    no_of_grid = x_size*y_size*z_size;
    grid_x = floor((point(1) - x_lower) / grid_size);
    grid_y = floor((point(2) - y_lower) / grid_size);
    grid_z = floor((point(3) - z_lower) / grid_size);

    % Calculate single grid ID
    gridID = grid_x + grid_y * ((x_upper - x_lower) / grid_size) + ...
        grid_z * ((x_upper - x_lower) / grid_size ) * ((y_upper - y_lower) / grid_size) + 1;

    if gridID < 0
        GIGA_ERROR = 1;
    end
    if find_neighbor
        neighour_list = [gridID+1; gridID-1; gridID+x_size; gridID+x_size+1; gridID+x_size-1; gridID-x_size; gridID-x_size+1; ... 
            gridID-x_size-1; gridID+(x_size*y_size); gridID+(x_size*y_size)+1; gridID+(x_size*y_size)-1; ...
            gridID+(x_size*y_size)+x_size; gridID+(x_size*y_size)+x_size+1; gridID+(x_size*y_size)+x_size-1; ... 
            gridID+(x_size*y_size)-x_size; gridID+(x_size*y_size)-x_size+1; gridID+(x_size*y_size)-x_size-1;... 
            gridID-(x_size*y_size); gridID-(x_size*y_size)+1; gridID-(x_size*y_size)-1; ... 
            gridID-(x_size*y_size)+x_size; gridID-(x_size*y_size)+x_size+1; gridID-(x_size*y_size)+x_size-1; ... 
            gridID-(x_size*y_size)-x_size; gridID-(x_size*y_size)-x_size+1; gridID-(x_size*y_size)-x_size-1];
        min_check = neighour_list < 0;
        max_check = neighour_list > no_of_grid;
        check = (min_check+max_check)>0;
        neighour_list(check) = [];
    end
    else
        grid_x = floor((point(1) - x_lower) / grid_size);
        grid_y = floor((point(2) - y_lower) / grid_size);
    
        % Calculate single grid ID
        gridID = grid_x + grid_y * ((x_upper - x_lower) / grid_size) + 1;
        if find_neighbor
            if grid_x == 0 % No Left neighbour
                neighour_list = [neighour_list; gridID+1];
            elseif grid_x == ((x_upper - x_lower) / grid_size)-1 % No Right neighbour
                neighour_list = [neighour_list; gridID-1];
            else % HAVE BOTH left and right neighbour
                neighour_list = [neighour_list; gridID-1; gridID+1];
            end

            if grid_y == 0 % NO down neighbour
                tempID = grid_x + (grid_y+1) * ((x_upper - x_lower) / grid_size) + 1;
                neighour_list = [neighour_list; tempID];
                    if grid_x == 0 % No Left neighbour
                        neighour_list = [neighour_list; tempID+1];
                    elseif grid_x == ((x_upper - x_lower) / grid_size)-1 % No Right neighbour
                        neighour_list = [neighour_list; tempID-1];
                    else % HAVE BOTH left and right neighbour
                        neighour_list = [neighour_list; tempID-1; tempID+1];
                    end
            elseif grid_y == ((y_upper - y_lower) / grid_size)-1 % NO up neighbour
                tempID = grid_x + (grid_y-1) * ((x_upper - x_lower) / grid_size) + 1;
                neighour_list = [neighour_list; tempID];
                    if grid_x == 0 % No Left neighbour
                        neighour_list = [neighour_list; tempID+1];
                    elseif grid_x == ((x_upper - x_lower) / grid_size)-1 % No Right neighbour
                        neighour_list = [neighour_list; tempID-1];
                    else % HAVE BOTH left and right neighbour
                        neighour_list = [neighour_list; tempID-1; tempID+1];
                    end
            else
                tempID1 = grid_x + (grid_y-1) * ((x_upper - x_lower) / grid_size) + 1; % UP neighbour
                tempID2 = grid_x + (grid_y+1) * ((x_upper - x_lower) / grid_size) + 1; % DOWN neighour
                neighour_list = [neighour_list; tempID1; tempID2];
                    if grid_x == 0 % No Left neighbour
                        neighour_list = [neighour_list; tempID1+1; tempID2+1];
                    elseif grid_x == ((x_upper - x_lower) / grid_size)-1 % No Right neighbour
                        neighour_list = [neighour_list; tempID1-1; tempID2-1];
                    else % HAVE BOTH left and right neighbour
                        neighour_list = [neighour_list; tempID1-1; tempID1+1; tempID2-1; tempID2+1];
                    end
            end
        end
    end
end

File: get_grid_id.py
import numpy as np


def get_grid_id(point, grid_size, x_lower, x_upper, y_lower, y_upper,
                z_lower, z_upper, dimension, find_neighbor):
    """
    Compute grid ID for a 3D or 2D point, and optionally its neighbour list.

    Parameters
    ----------
    point : np.ndarray (3,) or (2,)
        Point coordinates.
    grid_size : float
        Grid cell size.
    x_lower, x_upper : float
        X-axis bounds.
    y_lower, y_upper : float
        Y-axis bounds.
    z_lower, z_upper : float
        Z-axis bounds (ignored for 2D).
    dimension : int
        2 or 3.
    find_neighbor : bool
        Whether to compute neighbour list.

    Returns
    -------
    grid_id : int
        1-based grid index.
    neighbour_list : list
        List of neighbouring grid IDs (1-based). Only returned if find_neighbor=True.
    """
    neighbour_list = []

    if dimension == 3:
        point = point.copy()
        if point[2] < 0:
            point[2] = 0

        x_size = (x_upper - x_lower) / grid_size
        y_size = (y_upper - y_lower) / grid_size
        z_size = (z_upper - z_lower) / grid_size
        no_of_grid = int(x_size * y_size * z_size)

        grid_x = int(np.floor((point[0] - x_lower) / grid_size))
        grid_y = int(np.floor((point[1] - y_lower) / grid_size))
        grid_z = int(np.floor((point[2] - z_lower) / grid_size))

        grid_id = grid_x + grid_y * int(x_size) + grid_z * int(x_size) * int(y_size) + 1

        if grid_id < 0:
            raise ValueError("grid_id < 0")

        if find_neighbor:
            xs = int(x_size)
            ys = int(y_size)
            candidates = [
                grid_id + 1, grid_id - 1,
                grid_id + xs, grid_id + xs + 1, grid_id + xs - 1,
                grid_id - xs, grid_id - xs + 1, grid_id - xs - 1,
                grid_id + xs * ys,
                grid_id + xs * ys + 1,
                grid_id + xs * ys - 1,
                grid_id + xs * ys + xs,
                grid_id + xs * ys + xs + 1,
                grid_id + xs * ys + xs - 1,
                grid_id + xs * ys - xs,
                grid_id + xs * ys - xs + 1,
                grid_id + xs * ys - xs - 1,
                grid_id - xs * ys,
                grid_id - xs * ys + 1,
                grid_id - xs * ys - 1,
                grid_id - xs * ys + xs,
                grid_id - xs * ys + xs + 1,
                grid_id - xs * ys + xs - 1,
                grid_id - xs * ys - xs,
                grid_id - xs * ys - xs + 1,
                grid_id - xs * ys - xs - 1,
            ]
            neighbour_list = [n for n in candidates if 1 <= n <= no_of_grid]

    else:  # dimension == 2
        x_size = (x_upper - x_lower) / grid_size
        y_size = (y_upper - y_lower) / grid_size
        xs = int(x_size)

        grid_x = int(np.floor((point[0] - x_lower) / grid_size))
        grid_y = int(np.floor((point[1] - y_lower) / grid_size))
        grid_id = grid_x + grid_y * xs + 1

        if find_neighbor:
            # Left/Right neighbours
            if grid_x == 0:
                neighbour_list.append(grid_id + 1)
            elif grid_x == xs - 1:
                neighbour_list.append(grid_id - 1)
            else:
                neighbour_list.extend([grid_id - 1, grid_id + 1])

            if grid_y == 0:
                temp_id = grid_x + (grid_y + 1) * xs + 1
                neighbour_list.append(temp_id)
                if grid_x == 0:
                    neighbour_list.append(temp_id + 1)
                elif grid_x == xs - 1:
                    neighbour_list.append(temp_id - 1)
                else:
                    neighbour_list.extend([temp_id - 1, temp_id + 1])
            elif grid_y == int(y_size) - 1:
                temp_id = grid_x + (grid_y - 1) * xs + 1
                neighbour_list.append(temp_id)
                if grid_x == 0:
                    neighbour_list.append(temp_id + 1)
                elif grid_x == xs - 1:
                    neighbour_list.append(temp_id - 1)
                else:
                    neighbour_list.extend([temp_id - 1, temp_id + 1])
            else:
                temp_id1 = grid_x + (grid_y - 1) * xs + 1  # UP
                temp_id2 = grid_x + (grid_y + 1) * xs + 1  # DOWN
                neighbour_list.extend([temp_id1, temp_id2])
                if grid_x == 0:
                    neighbour_list.extend([temp_id1 + 1, temp_id2 + 1])
                elif grid_x == xs - 1:
                    neighbour_list.extend([temp_id1 - 1, temp_id2 - 1])
                else:
                    neighbour_list.extend([
                        temp_id1 - 1, temp_id1 + 1,
                        temp_id2 - 1, temp_id2 + 1,
                    ])

    return grid_id, neighbour_list
